modcc/functioninliner.cpp

```diff
@@ -1,85 +1,157 @@
 #include <iostream>
 
+#include "astmanip.hpp"
 #include "error.hpp"
 #include "functioninliner.hpp"
 #include "errorvisitor.hpp"
 
-expression_ptr inline_function_call(Expression* e)
+expression_ptr inline_function_call(const expression_ptr& e)
 {
-    if(auto f=e->is_function_call()) {
-        auto func = f->function();
-#ifdef LOGGING
-        std::cout << "inline_function_call for statement " << f->to_string()
-                  << " with body" << func->body()->to_string() << "\n";
-#endif
-        auto& body = func->body()->statements();
-        if(body.size() != 1) {
-            throw compiler_exception(
-                "can only inline functions with one statement", func->location()
-            );
+    auto assign_to_func = e->is_assignment();
+    auto ret_name = assign_to_func->lhs()->is_identifier()->clone();
+
+    if(auto f = assign_to_func->rhs()->is_function_call()) {
+        auto body = f->function()->body()->clone();
+
+        for (auto&s: body->is_block()->statements()) {
+            s->semantic(e->scope());
         }
 
-        if(body.front()->is_if()) {
-            throw compiler_exception(
-                    "can not inline functions with if statements", func->location()
-            );
+        FunctionInliner func_inliner(f->name(), ret_name, f->function()->args(), f->args(), e->scope());
+
+        body->accept(&func_inliner);
+        if (!func_inliner.return_val_set()) {
+            throw compiler_exception(pprintf("return variable of function % not set", f->name()), e->location());
         }
+        return body;
+    }
+    return {};
+}
+///////////////////////////////////////////////////////////////////////////////
+//  function inliner
+///////////////////////////////////////////////////////////////////////////////
+void FunctionInliner::replace_with_args(Expression* e) {
```

Comment:
Please add a comment above this function definition that describes what it does.

Reply:
Done